man/upsconfig.8
.TH UPSCONFIG 8 "Wed Aug 23 2006" "" "Network UPS Tools (NUT)" 
.SH NAME
upsconfig \- simple UPS configuration tool
.SH SYNOPSIS
.B upsconfig \-h

.B upsconfig \-g <varname>

.B upsconfig \-d <driver> [\-m <mode>] [\-p <port>] [\-t <target_dir>] [\-b <base_config_dir>] [\-u <upsd_server>] [\-s] [\-h]

.SH DESCRIPTION

.B upseconfig
is to be used to configure NUT (version 2.1 or higher) for simple use case.
That is to say only one UPS. It could be used for more complexe case in
order to make a "base" and to further modify it.

.SH OPTIONS

.IP "\-b \fItemplate_dir\fR, \-\-base_config_dir \fItemplate_dir\fR"
Specify the directory where \fBupsconfig\fR should find the configuration template
to use to generate a new config. Note that this directory should contains a
directory named comments, containing the comments templates for your configuration
to be commented.

.IP "\-d \fIdriver\fR, \-\-driver \fIdriver\fR"
The driver which support your UPS. this option is mandatory

.IP "\-g \fIvarname\fR, \-\-get \fIvarname\fR"
Return the value of the variable \fIvarname\fR in the current configuration file.
Return Undefined if this variable doesn't exist or doesn't has value.

.IP "\-h, \-\-help"
Display the help text.

.IP "\-m \fImode\fR, \-\-mode \fImode\fR"
Specify the mode of functionement of NUT. This value was not specified in
the old format. Valid value are :
.IP
.nf
	none       \(hy NUT is not configured. This is the default value
	standalone \(hy A computer protected by UPS, without client accessing to the upsd server by network.
	net_server \(hy This computer run the upsd server so that other computers can acces to it by network.
	net_client \(hy This computer relies on another computer in the network to run upsd.
	pm         \(hy NUT is used by power management.
.fi
.LP

.IP "\-p \fIport\fR, \-\-port \fIport\fR"
The port on which your UPS is connected. Mandatory for non USB UPS

.IP "\-q, \-\-quiet"
\fBmigrateconfig\fR will not print anything other than errors.

.IP "\-s, \-\-single"
\fBupsconfig\fR will save the configuration in a single file : nut.conf.
If not specified, \fBupsconfig\fR will use "include" directive and let
the configuration splited in four files like with the old format (at the exception that
"upsd.users" has been renamed "users.conf")

.IP "\-t \fIdestination\fR, \-\-target_dir \fIdestination\fR"
Specify the directory where \fBmigrateconfig\fR will save the configuration
in the new format.

.IP "\-u \fIserver\fR, \-\-upsd_server \fIserver\fR"
Specify where run the upsd server in net_client mode. Must be given in the standard address:port
format. (example : foobox:1234)

.SH WARNING

Be sure to verify the output configuration file before using it. Especially,
You will need to modify password of the users in order for NUT to accept them (\fBupsconfig\fR set the 
password at '!', which make NUT to reject the whole user). In net_server mode, you will want to
modify/add acl and accept/reject variable in the upsd section in order to allow the good computers
to connect to your upsd server.

To prevent confusion with old files, \fBupsconfig\fR save any configuration file in the target directory
by adding ".sav" to its name. If files having this name already exist, \fBmigrateconfig\fR exits without
doing the translation (but some old configuration files may already have been renamed). This apply to files named :
nut.conf, ups.conf, users.conf, upsd.conf and upsmon.conf

.SH SEE ALSO

.SS Old format configuration files:
\fBups.conf\fR(5), \fBupsd.conf\fR(5), 
\fBupsd.users\fR(5), \fBupsmon.conf\fR(5)

.SS New format configuration file:
\fBnut.conf\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


CHANGES
Thu Aug 31 15:07:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - drivers/mains.c : some bugs corrected
 - lib/libupsconfig.c : now allow, when a list of string is expected for a variable, to pu a string
   this string will be considered as a list of string with only one element. this will probably prevent
   a lot of wondering from the users, being more flexible. ( and with only one element, it is perhaps more
   natural to put something like allowfrom = "localhost" than allowfrom = { "localhost" }. Moreover, the
   users won't have to know if the variable need a list or not : if they have only one value, they can alway use string,
   if they have more than one, it is obvious it is a list of string that is needed ^_^

Thu Aug 31 11:31:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - lib/libupsconfig.ch : added the function get_comments_templates
 - utils/upsconfig.c : modified to use the function get_comments_templates
 - utils/migrateconfig.c : modified to use the function get_comments_templates

Thu Aug 31 90:57:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - lib/libupsconfig.c : Arnaud defined what need to be saved in configuration file
   when in mode "pm". Modified the code accordingly.

Thu Aug 31 09:25:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - clients/upsmon.c : correction of a little bug
 - lib/libupsconfig.ch : function that returns integer now return -1 if they failed, not 0 (because 0 can be a valide value)
 - utils/migrateconfig.c : some minor modifications
 - utils/upsconfig.c : added the -g option, to get the value of a variable in the configuration file. (for instance to get the powerdownflag in shutdown script)
 - man/upsconfig.8 : added the -g option
 - server/user.c : correction of some bugs

Tue Aug 29 10:11:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - lib/libupsconfig.ch : disabled the saving in configuration
   file of the notion of rights. For more info, look just below
   the header of libupsconfig.h
 - conf/comments/conf.comm.C : removed references to the rights notion
   save of the file with those reference has been saved and will be put
   on alioth for future when it will be enable again
 - man/nut.conf.5 : same comment than for conf.comm.C

Fri Aug 25 10:54:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - man/nut.conf : added
 - man/upsd.8
      /upsmon.8     Modified for the new configuration
      /nutupsdrv.8
 - utils/migrateconfig : Modified to handle multiple accept reject occurence
 - server/conf.c : Modified to take into account reject then accept
 - server/user.c : added actions FSD by default for upsmon user (master and slave)

Thu Aug 24 14:54:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - Many files I created : cleaning to replace // comment by /* */ comments
 - drivers/main.c upsdrvctl : global variable for ups.conf are now in nut.ups.global.
 - lib/libupsconfig.c : When saving, only try to save existant configuration files 
   (not all configuration files possible XD)
 - utils : renaming migratetool into migrateconfig
 - man : added manual for migrateconfig and upsconfig

Wed Aug 9  17:06:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - utils/migratetool.c : A tool to migrate from old config format to the new one
 - utils/Makefile.in : added the rules to compile migratetool
 - lib/libupsconfig.c : little to prevent a segfault in case the RUN_AS_USER group don't exist
 - common/datatypes.c : minor changes
 - utils/upsconfig.c : minor changes

Fri Aug 4  15:44:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - client/upsmon.c : adaptated to use libupsconfig instead of parseconf
 - drivers/main.ch, upsdrvctl.c : adaptated to use libupsconfig
 - server/access.ch, conf.c, user.ch : adapted to use libupsconfig
 - Makefile.in in the client, driver and server directory. gendb in drivers directory
   Adapted to use libupsconfig instead of parseconf or upsconf
 - lib/makefile : adapted to make the libupsconfig.a library
 - lib/libupsconfig.ch : added some functions
 - common/data_types.c : corrected a omission
 - data/base_config/nut.conf : Some minor changes
 - utils/Makefile.in : adapted to use libupsconfig as a library
 - utils/upsconfig.c : a minor change
 - Makefile : Changed the order of the subdir to make libupsconfig.a before what need it

Tue Aug 1  17:06:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - utils/upsconfig.c : fully functionnal and commented version of upsconfig !!
 - lib/libupsconfig.h : fully commented
 - lib/libupsconfig.c : some minor fixes
 - include/data_type.h stack.h tree.h : added comments

Mon Jul 31 16:17:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - common/data_types.ch : added some conversion functions
 - nutparser.ch : some minor changes
 - tree.ch : functions do not anymore exit on error but return an error value
 - data/base_config : a new directory containing template to generate configuration files
 - lib/libupsconfig.ch : lot of little changes. functions do not anymore exit on error 
   but return an error value.
 - utils/upsconfig.c : creation of this tool.
   * utils/Makefile.in added too.
 - configure.in : added a rule to make utils/Makefile
 - Makefile.in : added the utils directory in the SUBDIR variable

Wed Jul 26 17:03:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - data_types.ch : moving in some functions that should be here and not elsewhere
 - tree.ch : moving in a function that should be here and not elsewhere
 - libupsconfig.ch : nearly finished. Have still some test to do.
 - Makefile.in (at the root) : added lib in the list of SUBDIR
 - lib/Makefile.in : added rules to compile libupsconfig

Mon Jul 24 17:00:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - test/*.conf : added, to allow testing the parser

Fri Jul 21 16:34:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - Makefile.in : added test in the SUBDIR variable to make also test_parser
 

utils/upsconfig.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <pwd.h> 

#include "data_types.h"
#include "common.h"
#include "../lib/libupsconfig.h"
#include "nutparser.h"


void print_error(const char* errtxt) {
	fprintf(stderr, "Upsconfig error : %s. Aborting\n", errtxt);
}

void print_usage() {
	printf("usage : upsconfig -d <driver> [-m <mode>] [-p <port>] [-t <target_dir>]\n");
	printf("                  [-b <base_config_dir>] [-u <upsd_server>] [-s]\n");
	printf("        upsconfig -g <variable name>\n");
	printf("        upsconfig -h\n");
	printf("-d, --driver            The driver which support the ups\n");
	printf("-m, --mode              The mode to run NUT in. standalone | net_server |\n");
	printf("                        net_client | pm | none.\n");
	printf("-p, --port              The port in which the ups is plugged. Needed if the\n");
	printf("                        driver don't support \"auto\". default is \"auto\"\n");
	printf("-t, --target_dir        Where to save the generated configuration. Default\n");
	printf("                        is %s\n", CONFPATH);
	printf("-b, --base_config_dir   The directory where are the templates and comments\n");
	printf("                        files. Default is %s\n", CONFPATH);
	printf("-u, --upsd_server       Where is the upsd server, in host[:port] format\n");
	printf("                        Use it for net_client configuration. Default is\n");
	printf("                        \"localhost\"\n");
	printf("-g, --get               get the value of the given variable in configuration file\n");
	printf("-s, --single            Save the configuration in a single file\n");
	printf("-q, --quiet             Don't print anything apart errors\n");
	printf("-h, --help              Show this help message\n");
}



int main (int argc, char** argv)  {
	t_modes mode = standalone;
	t_string driver = 0;
	t_string port = "auto";
	t_string target_dir = 0;
	t_string base_config_dir;
	t_string comments_dir;
	t_string server = "localhost";
	t_string variable = 0;
	boolean single = FALSE;
	boolean quiet = FALSE;
	boolean get = FALSE;
	t_typed_value value;
	int i;
	t_string conf_file, comm_file, s;
	
	
	/* Value by default */
	target_dir = CONFPATH;
	base_config_dir = xmalloc(sizeof(char) * (strlen(CONFPATH) + 13));
	sprintf(base_config_dir, "%s/base_config", CONFPATH);
	
	
	
	/* Parse the parameters */
	for (i = 1; i <= (argc - 1); i++ ) {
		if (strcmp(argv[i], "-m") == 0 || strcmp(argv[i], "--mode") == 0) {
			if (i != argc - 1) {
				mode = string_to_mode(argv[++i]);
				if (mode == (unsigned int)-1) {
					print_error("bad parameter of option \"-m\"");
					print_usage();
					exit(EXIT_FAILURE);
				}
			} else {
				print_error("\"mode\" option used without parameter");
				print_usage();
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if (strcmp(argv[i], "-d") == 0 || strcmp(argv[i], "--driver") == 0) {
			if (i != argc - 1) {
				driver = string_copy(argv[++i]);
			} else {
				print_error("\"driver\" option used without parameter");
				print_usage();
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if (strcmp(argv[i], "-p") == 0 || strcmp(argv[i], "--port") == 0) {
			if (i != argc - 1) {
				port = string_copy(argv[++i]);
			} else {
				print_error("\"port\" option used without parameter");
				print_usage();
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if (strcmp(argv[i], "-t") == 0 || strcmp(argv[i], "--target_dir") == 0) {
			if (i != argc - 1) {
				target_dir = string_copy(argv[++i]);
			} else {
				print_error("\"target_dir\" option used without parameter");
				print_usage();
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if (strcmp(argv[i], "-b") == 0 || strcmp(argv[i], "--base_config_dir") == 0) {
			if (i != argc - 1) {
				free(base_config_dir);
				base_config_dir = string_copy(argv[++i]);
			} else {
				print_error("\"base_config_dir\" option used without parameter");
				print_usage();
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if (strcmp(argv[i], "-u") == 0 || strcmp(argv[i], "--upds_server") == 0) {
			if (i != argc - 1) {
				server = string_copy(argv[++i]);
			} else {
				print_error("\"upsd_server\" option used without parameter");
				print_usage();
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if (strcmp(argv[i], "-g") == 0 || strcmp(argv[i], "--get") == 0) {
			if (i != argc - 1) {
				variable = string_copy(argv[++i]);
				get = TRUE;
			} else {
				print_error("\"get\" option used without parameter");
				print_usage();
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if (strcmp(argv[i], "-s") == 0 || strcmp(argv[i], "--single") == 0) {
			single = TRUE;
			continue;
		}
		if (strcmp(argv[i], "-q") == 0 || strcmp(argv[i], "--quiet") == 0) {
			quiet = TRUE;
			continue;
		}
		if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
			print_usage();
			exit(0);
		}
	}
	
	if (get) {
		conf_file = xmalloc(sizeof(char) * (strlen(CONFPATH) + 10));
		sprintf(conf_file, "%s/nut.conf", CONFPATH);
	} else {
		/* Some parameters are mandatory */
		if (driver == 0 && mode != pm && mode != no_mode && mode != net_client) {
			print_error("\"driver\" option is mandatory");
			print_usage();
			exit(EXIT_FAILURE);
		}
	
		if (mode == net_client && server == 0) {
			print_error("\"upsd_server\" option is mandatory when in net_client mode");
			print_usage();
			exit(EXIT_FAILURE);
		}
		
		/* Generate the name of the template file to use */
		conf_file = xmalloc(sizeof(char) * (strlen(base_config_dir) + 10));
		sprintf(conf_file, "%s/nut.conf", base_config_dir);
	
		/* Generate the name of the comments file to use */	
		comments_dir = xmalloc(sizeof(char) * (strlen(base_config_dir) + 10));
		sprintf(comments_dir, "%s/comments", base_config_dir);
		comm_file = get_comments_template(comments_dir);
		

		free(base_config_dir);

		if (!quiet) {
			printf("\nLoading the base configuration from template %s\n\n", conf_file);
		}
	}

	/* Load the configuration */	
	if (!load_config(conf_file, 0)) {
		/* An errors occured, aborting */
		exit(EXIT_FAILURE);
	}
	
	if (get) {
		value = get_variable(variable);
		if (!value.has_value) {
			printf("Undefined\n");
			drop_config();
			exit(EXIT_SUCCESS);
		}
		if (value.type == string_type) {
			printf("%s\n", value.value.string_value); 
			drop_config();
			exit(EXIT_SUCCESS);
		}
		if (value.type == enum_string_type) {
			printf("%s\n", enum_string_to_string(value.value.enum_string_value)); 
		}
		drop_config();
		exit(EXIT_SUCCESS);
	}
	
	/* Modify the template configuration */
	
	/* Set the mode */
	set_mode(mode);
	
	/* Fill the ups section with the good values */
	search_ups("myups");
	set_driver(driver);
	set_port(port);
	
	set_run_as_user(RUN_AS_USER);
	
	if (mode == net_client) {
		/* net_client configuration don't need some part of the tree */
		remove_user("nutadmin");
		remove_user("monmaster");
		search_user("monslave");
		/* To remove the allowfrom variable : */
		set_allowfrom(0);
		/* if an host has been specified, should modify the host of the monitor rule. 
		   Create the one wanted then delete the first */		
		search_monitor_rule(1);
		s = get_monitor_ups();
		add_monitor_rule(s, server, get_monitor_powervalue(), "monslave");
		free(s);
		if ( strcmp(server, "localhost") != 0) {
			remove_monitor_rule(1);
		}
	}
	
	if (mode == standalone) {
		remove_user("monslave");
	}
	
	
	/* Save the generated configuration */
	if (!quiet) {
		printf("Saving your configuration in %s\n", target_dir);
	
		if (comm_file == 0) {
			printf("I did not find any comments template file. Your configuration files will not be commented\n\n");
		} else {
			printf("I'll use the following comments template file to comment your configuration file : %s\n\n", comm_file);
		}
	}
	
	if (!save_config(target_dir, comm_file, single, print_error)) {
		exit(EXIT_FAILURE);
	}
	
	if (!quiet) {
		/* Print some comments about the configuration */
		if (mode == net_server) {
			if (single) {
				printf("#   You are in network server configuration\n\n");
				printf("# Don't forget to modify the acl (in upsd.acl section),\n");
				printf("# the accept and reject variable (both in upsd section)\n");
				printf("# and the allowfrom variable of user monslave (in \n");
				printf("# users.monslave section) to allow him to acces to the.\n");
				printf("# upsd server.\n\n");
			} else {
				printf("#   You are in network server configuration\n\n");
				printf("# Don't forget to modify the acl (in upsd.conf file),\n");
				printf("# the accept and reject variable (also in upsd.conf file)\n");
				printf("# and the allowfrom variable of user monslave (in \n");
				printf("# users.conf, monslave section) to allow him to acces to.\n");
				printf("# the upsd server.\n\n");
			}
		}
	
		if (getuid() != getpwnam("root")->pw_uid) {
			printf("#   You are not root\n");
			printf("# I am not able to modify the ownership and the group of the \n");
			printf("# created file(s).\n\n");
		}
		
		if (single) {
			printf("# !!                                                    #\n");
			printf("# !! DON'T FORGET TO MODIFY PASSWORDS IN USERS SECTION  #\n");
			printf("# !!                                                    #\n");
		} else {
			printf("# !!                                                      #\n");
			printf("# !! DON'T FORGET TO MODIFY PASSWORDS IN users.conf FILE  #\n");
			printf("# !!                                                      #\n");
		}
	}
	
	drop_config();
	
	return 0;
}
